icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -275,14 +297,19 @@ int32_t FormattedStringBuilder::prepareForInsertHelper(int32_t index, int32_t co
     int32_t oldZero = fZero;
     char16_t *oldChars = getCharPtr();
     Field *oldFields = getFieldPtr();
+    int32_t newLength;
+    if (uprv_add32_overflow(fLength, count, &newLength)) {
+        status = U_INPUT_TOO_LONG_ERROR;
+        return -1;
+    }
     if (fLength + count > oldCapacity) {
-        if ((fLength + count) > INT32_MAX / 2) {
-            // If we continue, then newCapacity will overflow int32_t in the next line.
+        int32_t newCapacity;
+        //  newCapacity = (fLength + count) * 2 w/ overflow checking.
+        if (uprv_add32_overflow(newLength, newLength, &newCapacity)) {
             status = U_INPUT_TOO_LONG_ERROR;
             return -1;
         }
```

Comment:
This is too strict. Both the old and new code fail when twice the newLength overflows, even though we can satisfy the current request with just newLength alone, which we now know didn't overflow. The exponential over-allocation is just an optimization, to reduce the number of memory allocations.
Suggestion:
```suggestion
        int32_t newCapacity;
        if (newLength <= INT32_MAX / 2) {
            newCapacity = newLength * 2;
        } else if (newLength <= INT32_MAX - 1000) {
            newCapacity = newLength + 1000;
        } else {
            newCapacity = INT32_MAX;
        }
```

Reply:
done

Reply:
somehow this relax of memory allocation size (for the case newLength > INT32_MAX / 2) will cause crash inside the first uprv_memcpy2 bewlow while running "format/ListFormatterTest/TestInt32Overflow -e" test. I have to revert back that change to the prior restriction to avoid create new problem.